rfcs/45-stageless.md

```diff
@@ -0,0 +1,537 @@
+  - competing read-locks caused by other systems will still need to complete before a data store can be mutated; the effect is to "ignore the locks of `A`", not "skip the check completely"
+
+In addition to their use in run criteria, atomic ordering constraints are extremely useful for forcing tight interlocking behavior between systems.
+They allow us to ensure that the state read from and written to the earlier system cannot be invalidated.
+This functionality is extremely useful when updating indexes (used to look up components by value):
+allowing you to ensure that the index is still valid when the system using the index uses it.
+
+Of course, atomic ordering constraints should be used thoughtfully.
+As the strictest of the three types of system ordering dependency, they can easily result in unsatisfiable schedules if applied to large groups of systems at once.
+
+### States
+
+**States** are one particularly common and powerful form of run criteria, allowing you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three sets of systems:
+
+1. **Update systems:** these systems run each frame if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(my_system.in_state(GameState::Playing))`
+2. **On-enter systems:** these systems run once when the specified state is entered.
+   1. `app.add_system(my_system.on_enter(GameState::Playing))`
+3. **On-exit systems:** these systems run once when the specified stated is exited.
+   1. `app.add_system(my_system.on_enter(GameState::Playing))`
+
+Update systems are by far the simplest: they're simply powered by run criteria.
+`.in_state` is precisely equivalent to `run_if_resource_is`, except with an additional trait bound that the resource must implement `State`.
+
+On-enter and on-exit systems are stored in dedicated schedules, two per state, within the `App's` `Schedules`.
+These schedules can be configured in all of the ordinary ways, but, as they live in different schedules, ordering cannot be defined relative to systems in the main schedule.
+
+Due to their disruptive and far-reaching effects, state transitions do not occur immediately.
+Instead, they are deferred (like commands), until the next `flush_state<S: State>` exclusive system runs.
+This system first runs the `on_exit` schedule of the previous state on the world, then runs the `on_enter` schedule of the new state on the world.
+Once that is complete, the exclusive system ends and control flow resumes as normal.

```

Comment:
I haven't seen usages of this functionality (arbitrarily many state changes in a single-frame / state stack) in practice. It adds a ton of complexity, so I decided to cut it from this RFC.

It should be mentioned in drawbacks though.

Reply:
The main reason for dropping the stack isn't technical limitations. The stack just wasn't working like users expected (e.g. <https://github.com/bevyengine/bevy/pull/2211#issuecomment-1002353919>) and we think it's a bit overkill now.

> If one wanted to emulate the current looping behavior in this new model, how would one do that?

You'd make your own state stack resource and implement your own `flush_state` exclusive system that, instead of running just one pair of exit-enter transitions, loops and runs transitions until there are none left. That would allow arbitrarily many state changes in a single frame.

If users request it, we can revisit making that the canonical states implementation again, but given how simple an extension that is, it's easier for us if we don't have to add more transition types (`on_pause`, `on_resume`, etc.).

Reply:
Added more clarity on this in drawbacks.

Reply:
I wasn't talking about a stack. That's orthogonal. I was talking about the looping model. I was talking about repeated state transitions in the same frame. We had that even before we had the stack. `StateStage` in 0.4 supported it. It was what motivated the looping stuff. Cart seemed very insistent that this should be supported.

Reply:
Multiple state transitions in a single schedule iteration would still be possible. Users can insert more than one instance of `flush_state` into a given schedule. Infinite transitions would not be possible because `flush_state` itself wouldn't loop.

It is a little arbitrary, but doing things like making one `on_enter` schedule trigger another, immediate state transition (in the same FSM) is also kinda weird.